### PR TITLE
fix: defer callback to prevent triggering resizeobserver loop limit

### PR DESCRIPTION
**Pull request by cyjake**
same issue as https://github.com/react-component/align/pull/72

### CODE REVIEW

File: tests/index.spec.js [hunk 1]
```diff
@@ -117,6 +130,19 @@ describe('ResizeObserver', () => {
...
+    function Component(props) {
+      return <ResizeObserver onResize={props.onResize}>{!props.hide && <div />}</ResizeObserver>;
+    }
+    const wrapper = mount(<Component onResize={onResize} />);
+
+    wrapper.setProps({ hide: true });
+    wrapper.triggerResize();
+    jest.runAllTimers();
+    expect(onResize).not.toHaveBeenCalled();
+  });
```
Comment (cyjake) [inline tests/index.spec.js:144]: This test case can re-produce the removed target issue.

Comment (cyjake) [reply tests/index.spec.js:144]: In real scenario, assembling Tabs dynamically might raise this issue as well. Take https://yuque.com for example. When editing flowchart in yuque editor, selection multiple items by mouse (which changes selections drastically) causes tabs on the right to re-render, which in turn causes following error

image

File: src/index.tsx [hunk 2]
```diff
@@ -80,6 +80,7 @@ class ReactResizeObserver extends React.Component<ResizeObserverProps, ResizeObs
     const { onResize } = this.props;
 
     const target = entries[0].target as HTMLElement;
+    if (!document.contains(target)) return;
```
Comment (zombieJ) [inline src/index.tsx:83]: `rc-resize-observer` 和 `rc-align` 不太一样，底层库即便上层不包含也可能有需求。这个暂时先不加吧。

File: src/index.tsx [hunk 3]
```diff
@@ -98,10 +99,12 @@ class ReactResizeObserver extends React.Component<ResizeObserverProps, ResizeObs
       this.setState(size);
 
       if (onResize) {
-        onResize({
-          ...size,
-          offsetWidth,
-          offsetHeight,
+        requestAnimationFrame(() => {
```
Comment (zombieJ) [inline src/index.tsx:102]: 是否可以放到 promise 里？`requestAnimationFrame` 会导致依赖同步渲染的有闪动。

File: src/index.tsx (hunk 2 above)
Comment (cyjake) [reply src/index.tsx:83]: 噢，也可以，那就是 Tab 的 onResize 需要判断一下 target 是否还在 dom 里了，我先调整一下这个 pr

**Comment by afc163** [comment]
antd 的测试用例挂了：https://app.circleci.com/pipelines/github/ant-design/ant-design/16566/workflows/5dd74ee7-2543-4965-a465-07b920cbaebc/jobs/226261

**Comment by zombieJ** [comment]
我看一下

**Pull request merged**